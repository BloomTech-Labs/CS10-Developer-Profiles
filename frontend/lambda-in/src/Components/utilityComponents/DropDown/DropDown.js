import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import Typography from '@material-ui/core/Typography';
import ExpansionPanel from '@material-ui/core/ExpansionPanel';
import ExpansionPanelSummary from '@material-ui/core/ExpansionPanelSummary';
import ExpansionPanelDetails from '@material-ui/core/ExpansionPanelDetails';
import Icon from '@material-ui/core/Icon';

/**
 * A wrapper component for @material-ui <ExpansionPanel />
 *
 * @description A Dropdown component wrapper, which accepts any type of children.
 *
 * @function A functional component.
 * @prop {string} header - Text to be displayed when the Dropdown is collapsed.
 *
 * @example How to use it
 * <DropDown header="Header text to display here">
 */
const DropDown = ({ header, children }) => (
  <Fragment>
    <div className="panelMargin">
      <ExpansionPanel>
        <ExpansionPanelSummary expandIcon={<Icon>add_circle</Icon>}>
          <Typography>{header}</Typography>
        </ExpansionPanelSummary>
        <ExpansionPanelDetails>
          <div className="expansionPanel">{children}</div>
        </ExpansionPanelDetails>
      </ExpansionPanel>
    </div>
  </Fragment>
);

DropDown.propTypes = {
  header: PropTypes.string.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
    PropTypes.func,
  ]).isRequired,
};

export default DropDown;
