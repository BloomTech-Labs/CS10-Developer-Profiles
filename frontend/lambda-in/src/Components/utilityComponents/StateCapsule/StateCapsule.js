import React, { Component } from 'react';
import PropTypes from 'prop-types';

/**
 * Create a local state to manage an object's properties whilist is edited.
 *
 * @description
 */
class StateCapsule extends Component {
  constructor(props) {
    super(props);
    this.state = {
      ready: false,
    };
    this.removeItem = this.removeItem.bind(this);
    this.createItem = this.createItem.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleKeyPress = this.handleKeyPress.bind(this);
    // prettier-ignore
    this.mapAndResetKeysValues = object => Object.keys(object)
      .reduce((resetKeys, key) => {
        const valueType = Object.prototype.toString.call(object[key]);

        switch (valueType) {
          case '[object Array]':
            // eslint-disable-next-line no-param-reassign
            resetKeys[key] = [];
            // eslint-disable-next-line no-param-reassign
            resetKeys[`${key}_edit`] = '';
            break;
          default:
            // eslint-disable-next-line no-param-reassign
            resetKeys[key] = '';
        }

        return resetKeys;
      }, {});
  }

  componentDidMount() {
    this.resetState();
  }

  resetState() {
    const { schema, object } = this.props;
    const newState = {
      ...this.mapAndResetKeysValues(schema),
      ...object,
      ready: true,
    };

    // Set local state with object-properties and initialize its values.
    this.setState(newState);
  }

  /**
   * Sync local state with input field.
   */
  handleChange(event) {
    event.stopPropagation();

    const { id } = event.target;
    const details = id.split('-');
    const field = details[1];
    const { value } = event.target;

    this.setState({ [field]: value });
  }

  handleKeyPress(e) {
    e.stopPropagation();
    const { chips } = e.target.dataset;

    if (e.key === 'Enter' && chips) {
      e.preventDefault();

      const details = e.target.id.split('-');
      const { field } = e.target.dataset;
      const { value } = e.target;

      this.setState((prevState) => {
        const toUpdate = prevState[field];
        toUpdate.push(value);
        return { [field]: toUpdate, [details[1]]: '' };
      });
    }
  }

  createItem(field) {
    const newData = { ...this.state };
    delete newData.ready;

    const createEvent = new CustomEvent('onCreateItem', {
      bubbles: true,
      detail: {
        newData,
        field,
      },
    });
    return (e) => {
      e.stopPropagation();

      e.target.dispatchEvent(createEvent);
      this.resetState();
    };
  }

  removeItem(field, index) {
    const removeEvent = new CustomEvent('onDeleteItem', {
      bubbles: true,
      detail: { field, index },
    });

    return (e) => {
      e.stopPropagation();

      //
      e.target.dispatchEvent(removeEvent);

      const typeOfField = Object.prototype.toString.call(field);

      if (
        // prettier-ignore
        typeOfField === '[object String]'
        || typeOfField === '[object Number]'
        || typeOfField === '[object Boolean]'
      ) {
        this.setState((prevState) => {
          const toUpdate = prevState[field];
          toUpdate.splice(index, 1);
          return { [field]: toUpdate };
        });
      }
    };
  }

  render() {
    const { ready } = this.state;

    const stateCapsule = { ...this.state };
    delete stateCapsule.ready;

    const { children } = this.props;

    return ready ? (
      // eslint-disable-next-line jsx-a11y/no-static-element-interactions
      <div
        className="state-capsule"
        onChange={this.handleChange}
        onKeyPress={this.handleKeyPress}
      >
        {children({
          stateCapsule,
          removeItem: this.removeItem,
          createItem: this.createItem,
        })}
      </div>
    ) : null;
  }
}

StateCapsule.propTypes = {
  schema: PropTypes.shape({}).isRequired,
  object: PropTypes.shape({}).isRequired,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
    PropTypes.func,
  ]).isRequired,
};

export default StateCapsule;
