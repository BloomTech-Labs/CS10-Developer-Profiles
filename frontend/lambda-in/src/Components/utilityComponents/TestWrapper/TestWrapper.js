import React, { Component, Fragment } from 'react';

/**
 * A Functional stateless component wrapper
 *
 * @description Enzyme's 'shallow' method don't create an instance for
 * stateless functional componets
 * Whitout an instance is not possible to access the functional componet's
 * props.
 * Because enzyme's shallow generate an instance for class coponents, <TestWrapper/>
 * allow us to access any functional coponent passed as a children.
 * @see https://github.com/airbnb/enzyme/issues/1232#issuecomment-334510754
 */
class TestWrapper extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  /**
   * Catch children component error an log them in test-enviroment
   *
   * @description An Error bundary to catch potential children component's
   * Javascrip errors
   * @see https://reactjs.org/docs/error-boundaries.html
   * @param {object} error - Error ebject
   * @param {object} info - Component stack on error
   */
  componentDidCatch(error, info) {
    // eslint-disable-next-line no-console
    console.log('Error Bundary', error, info);
  }

  render() {
    // eslint-disable-next-line react/prop-types
    const { children } = this.props;
    return <Fragment>{children({ ...this.props })}</Fragment>;
  }
}

export default TestWrapper;
