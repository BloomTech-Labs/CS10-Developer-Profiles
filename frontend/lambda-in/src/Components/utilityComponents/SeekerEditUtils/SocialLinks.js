import React, { Fragment } from 'react';
import TextField from '@material-ui/core/TextField';
import DropDown from '../DropDown/DropDown';

// eslint-disable-next-line arrow-parens
const SocialLinks = props => {
  // eslint-disable-next-line react/prop-types
  const { userInfo } = props;

  return (
    <Fragment>
      <DropDown header="Social links">
        <div className="inputField">
          <TextField id="github" label="github" fullWidth value={userInfo.github} margin="normal" variant="outlined" />
        </div>

        <div className="inputField">
          <TextField
            id="linkedin"
            label="Linkedin"
            fullWidth
            value={userInfo.linkedin}
            margin="normal"
            variant="outlined"
          />
        </div>

        <div className="inputField">
          <TextField
            id="portfolio"
            label="Portfolio Website"
            fullWidth
            value={userInfo.portfolio}
            margin="normal"
            variant="outlined"
          />
        </div>
      </DropDown>
    </Fragment>
  );
};

export default SocialLinks;
