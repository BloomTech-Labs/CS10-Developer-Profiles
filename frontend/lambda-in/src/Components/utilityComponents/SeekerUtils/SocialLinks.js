import React, { Fragment } from 'react';
import DropDown from '../DropDown';
import TextField from '@material-ui/core/TextField';

const SocialLinks = props => {
  const { userInfo } = props;

  return (
    <Fragment>
      <DropDown header="Social links">
        <div className="inputField">
          <TextField id="github" label="github" fullWidth value={userInfo.github} margin="normal" variant="outlined" />
        </div>

        <div className="inputField">
          <TextField
            id="linkedin"
            label="Linkedin"
            fullWidth
            value={userInfo.linkedin}
            margin="normal"
            variant="outlined"
          />
        </div>

        <div className="inputField">
          <TextField
            id="portfolio"
            label="Portfolio Website"
            fullWidth
            value={userInfo.portfolio}
            margin="normal"
            variant="outlined"
          />
        </div>
      </DropDown>
    </Fragment>
  );
};

export default SocialLinks;
