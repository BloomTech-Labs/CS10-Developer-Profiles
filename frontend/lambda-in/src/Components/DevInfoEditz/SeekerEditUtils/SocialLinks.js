import TextField from '@material-ui/core/TextField';
import React from 'react';
import SEEKER_PROP_TYPES from '../../seekerPropTypes';
import DropDown from '../../utilityComponents/DropDown/DropDown';
import StateCapsule from '../../utilityComponents/StateCapsule/StateCapsule';

const SocialLinks = ({ userInfo }) => {
  const schema = {
    github: userInfo.github,
    linkedin: userInfo.linkedin,
    portfolio: userInfo.portfolio,
  };

  return (
    <StateCapsule schema={schema} object={schema}>
      {({ stateCapsule }) => (
        <DropDown header="Social links">
          <div className="inputField">
            <TextField
              id="edit-github"
              label="github"
              fullWidth
              value={stateCapsule.github}
              margin="normal"
              variant="outlined"
            />
          </div>

          <div className="inputField">
            <TextField
              id="edit-linkedin"
              label="Linkedin"
              fullWidth
              value={stateCapsule.linkedin}
              margin="normal"
              variant="outlined"
            />
          </div>

          <div className="inputField">
            <TextField
              id="edit-portfolio"
              label="Portfolio Website"
              fullWidth
              value={stateCapsule.portfolio}
              margin="normal"
              variant="outlined"
            />
          </div>
        </DropDown>
      )}
    </StateCapsule>
  );
};

SocialLinks.propTypes = {
  userInfo: SEEKER_PROP_TYPES.isRequired,
};

export default SocialLinks;
