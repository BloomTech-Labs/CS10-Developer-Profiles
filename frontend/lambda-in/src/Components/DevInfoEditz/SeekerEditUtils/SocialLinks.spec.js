import React from 'react';
import { configure, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import SocialLinks from './SocialLinks';
import TestWrapper from '../../utilityComponents/TestWrapper/TestWrapper';
import StateCapsule from '../../utilityComponents/StateCapsule/StateCapsule';
import DropDown from '../../utilityComponents/DropDown/DropDown';

configure({ adapter: new Adapter() });

/**
 * @author jesuarva
 * We are only testing proper rendering of inner components,
 * this is accomplish with a snapshot.
 *
 * @see StateCapsule.spec.js for specific test related to its fucntionality.
 * @see DropDown.spec.js for specific test related to its fucntionality.
 */
describe('SocialLinks component', () => {
  const userInfo = {
    github: 'http://github.devprofiles',
    linkedin: 'http://linkedin.devprofiles',
    portfolio: 'http://portfolio.devprofiles',
  };
  const socialLinksComponent = <SocialLinks userInfo={userInfo} />;

  it('should renders correctly', () => {
    const tree = renderer.create(socialLinksComponent);
    expect(tree).toMatchSnapshot();
  });

  describe('Renders children components', () => {
    const mounted = mount(
      <TestWrapper>{() => socialLinksComponent}</TestWrapper>,
    );

    it('should renders a SteceCapsule component', () => {
      const stateCapsule = mounted.find(StateCapsule);
      expect(stateCapsule).toHaveLength(1);
    });

    it('should renders a DropDown component', () => {
      const dropDown = mounted.find(DropDown);
      expect(dropDown).toHaveLength(1);
    });
  });
});
