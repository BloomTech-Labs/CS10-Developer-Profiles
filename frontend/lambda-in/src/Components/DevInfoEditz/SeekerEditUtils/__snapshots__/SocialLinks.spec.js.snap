// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SocialLinks component should renders correctly 1`] = `
<div
  className="state-capsule"
  onChange={[Function]}
  onKeyPress={[Function]}
>
  <div
    className="panelMargin"
  >
    <div
      className="MuiPaper-root-4 MuiPaper-elevation1-7 MuiExpansionPanel-root-1"
    >
      <div
        aria-expanded={false}
        className="MuiButtonBase-root-37 MuiExpansionPanelSummary-root-31"
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex="0"
      >
        <div
          className="MuiExpansionPanelSummary-content-35"
        >
          <p
            className="MuiTypography-root-40 MuiTypography-body1-49"
          >
            Social links
          </p>
        </div>
        <div
          aria-hidden="true"
          className="MuiButtonBase-root-37 MuiIconButton-root-66 MuiExpansionPanelSummary-expandIcon-36"
          onBlur={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={-1}
        >
          <span
            className="MuiIconButton-label-71"
          >
            <span
              aria-hidden="true"
              className="material-icons MuiIcon-root-72"
            >
              add_circle
            </span>
          </span>
          <span
            className="MuiTouchRipple-root-81"
          />
        </div>
      </div>
      <div
        aria-hidden="true"
        className="MuiCollapse-container-88"
        style={
          Object {
            "minHeight": "0px",
          }
        }
      >
        <div
          className="MuiCollapse-wrapper-90"
        >
          <div
            className="MuiCollapse-wrapperInner-91"
          >
            <div
              className="MuiExpansionPanelDetails-root-92"
            >
              <div
                className="expansionPanel"
              >
                <div
                  className="inputField"
                >
                  <div
                    aria-describedby={undefined}
                    className="MuiFormControl-root-93 MuiFormControl-marginNormal-94 MuiFormControl-fullWidth-96"
                  >
                    <label
                      className="MuiFormLabel-root-104 MuiFormLabel-filled-108 MuiInputLabel-root-97 MuiInputLabel-formControl-98 MuiInputLabel-animated-101 MuiInputLabel-shrink-100 MuiInputLabel-outlined-103"
                      data-shrink={true}
                      htmlFor="edit-github"
                    >
                      github
                    </label>
                    <div
                      className="MuiInputBase-root-124 MuiOutlinedInput-root-111 MuiInputBase-fullWidth-133 MuiInputBase-formControl-125"
                      onClick={[Function]}
                    >
                      <fieldset
                        aria-hidden={true}
                        className="MuiNotchedOutline-root-141 MuiOutlinedInput-notchedOutline-118"
                        style={
                          Object {
                            "paddingLeft": 8,
                          }
                        }
                      >
                        <legend
                          className="MuiNotchedOutline-legend-142"
                          style={
                            Object {
                              "width": NaN,
                            }
                          }
                        />
                      </fieldset>
                      <input
                        aria-invalid={false}
                        autoComplete={undefined}
                        autoFocus={undefined}
                        className="MuiInputBase-input-134 MuiOutlinedInput-input-119"
                        defaultValue={undefined}
                        disabled={false}
                        id="edit-github"
                        name={undefined}
                        onBlur={[Function]}
                        onChange={[Function]}
                        onFocus={[Function]}
                        onKeyDown={undefined}
                        onKeyUp={undefined}
                        placeholder={undefined}
                        readOnly={undefined}
                        required={false}
                        rows={undefined}
                        type="text"
                        value="https://github.com/talent_developer"
                      />
                    </div>
                  </div>
                </div>
                <div
                  className="inputField"
                >
                  <div
                    aria-describedby={undefined}
                    className="MuiFormControl-root-93 MuiFormControl-marginNormal-94 MuiFormControl-fullWidth-96"
                  >
                    <label
                      className="MuiFormLabel-root-104 MuiFormLabel-filled-108 MuiInputLabel-root-97 MuiInputLabel-formControl-98 MuiInputLabel-animated-101 MuiInputLabel-shrink-100 MuiInputLabel-outlined-103"
                      data-shrink={true}
                      htmlFor="edit-linkedin"
                    >
                      Linkedin
                    </label>
                    <div
                      className="MuiInputBase-root-124 MuiOutlinedInput-root-111 MuiInputBase-fullWidth-133 MuiInputBase-formControl-125"
                      onClick={[Function]}
                    >
                      <fieldset
                        aria-hidden={true}
                        className="MuiNotchedOutline-root-141 MuiOutlinedInput-notchedOutline-118"
                        style={
                          Object {
                            "paddingLeft": 8,
                          }
                        }
                      >
                        <legend
                          className="MuiNotchedOutline-legend-142"
                          style={
                            Object {
                              "width": NaN,
                            }
                          }
                        />
                      </fieldset>
                      <input
                        aria-invalid={false}
                        autoComplete={undefined}
                        autoFocus={undefined}
                        className="MuiInputBase-input-134 MuiOutlinedInput-input-119"
                        defaultValue={undefined}
                        disabled={false}
                        id="edit-linkedin"
                        name={undefined}
                        onBlur={[Function]}
                        onChange={[Function]}
                        onFocus={[Function]}
                        onKeyDown={undefined}
                        onKeyUp={undefined}
                        placeholder={undefined}
                        readOnly={undefined}
                        required={false}
                        rows={undefined}
                        type="text"
                        value="Linkedin/talent_developer"
                      />
                    </div>
                  </div>
                </div>
                <div
                  className="inputField"
                >
                  <div
                    aria-describedby={undefined}
                    className="MuiFormControl-root-93 MuiFormControl-marginNormal-94 MuiFormControl-fullWidth-96"
                  >
                    <label
                      className="MuiFormLabel-root-104 MuiFormLabel-filled-108 MuiInputLabel-root-97 MuiInputLabel-formControl-98 MuiInputLabel-animated-101 MuiInputLabel-shrink-100 MuiInputLabel-outlined-103"
                      data-shrink={true}
                      htmlFor="edit-portfolio"
                    >
                      Portfolio Website
                    </label>
                    <div
                      className="MuiInputBase-root-124 MuiOutlinedInput-root-111 MuiInputBase-fullWidth-133 MuiInputBase-formControl-125"
                      onClick={[Function]}
                    >
                      <fieldset
                        aria-hidden={true}
                        className="MuiNotchedOutline-root-141 MuiOutlinedInput-notchedOutline-118"
                        style={
                          Object {
                            "paddingLeft": 8,
                          }
                        }
                      >
                        <legend
                          className="MuiNotchedOutline-legend-142"
                          style={
                            Object {
                              "width": NaN,
                            }
                          }
                        />
                      </fieldset>
                      <input
                        aria-invalid={false}
                        autoComplete={undefined}
                        autoFocus={undefined}
                        className="MuiInputBase-input-134 MuiOutlinedInput-input-119"
                        defaultValue={undefined}
                        disabled={false}
                        id="edit-portfolio"
                        name={undefined}
                        onBlur={[Function]}
                        onChange={[Function]}
                        onFocus={[Function]}
                        onKeyDown={undefined}
                        onKeyUp={undefined}
                        placeholder={undefined}
                        readOnly={undefined}
                        required={false}
                        rows={undefined}
                        type="text"
                        value="portfolio.talent_developer"
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
